torch/fx/experimental/symbolic_shapes.py

```diff
@@ -254,10 +254,50 @@ def _sympystr(self, printer):
 
         @classmethod
         def eval(cls, base, divisor):
+            # NOTE [ Checking types with SymPy ]
+            # Python has a dedicated complex type, but SymPy represents complex
+            # with Add exprs.
+            #
+            # isinstance doesn't work for arbitrary exprs and for Symbols, even
+            # when the latter have integer=True set during construction. So you
+            # are supposed to use the is_* properties, but it's also not always
+            # reliable as 0 has both is_integer and is_real set to True no
+            # matter whether it's constructed as a Float or an Integer.
+            #
+            # Also, booleans cannot be used in arithmetic exprs in SymPy and
+            # cannot be passed to the Integer and Float constructors.
+            # https://github.com/pytorch/pytorch/issues/90900
```

Comment:
I guess my problem with this comment is that it reads like "well, I tried a bunch of things and they didn't work so here's the thing we're doing now" but the comment never lays out what it is exactly that you're trying to do, so an uniformed reader will look at the comment and be like, "ok, so sympy represents complex with add exprs, why do I care".

But Sympy in fact gives *explicit* guidance about what to do when defining functions restricted to certain domains, e.g., https://docs.sympy.org/latest/guides/assumptions.html#implementing-assumptions-handlers so really we should quote the docs and then follow what they say there. (BTW, it doesn't look like your code does what the docs say to do lol)

Reply:
Updated the comment and added things specific to the assumptions system.

I still need to rebase and clean up minor issues, but feedback welcome.